src/V3DynamicClass.cpp

```diff
@@ -0,0 +1,93 @@
+//
+//*************************************************************************
+//
+// Copyright 2003-2022 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3DynamicClass's Transformations:
+//
+//      Each class with an event member:
+//          In the destructor, notify the event dispatcher that the event var doesn't exist
+//          anymore.
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3DynamicClass.h"
+#include "V3Ast.h"
+
+//######################################################################
+// Manage the lifetime of class member events
+
+class DynamicClassEventVisitor final : public VNVisitor {
+private:
+    // STATE
+    AstClass* m_classp = nullptr;  // Current class
+    AstNode* m_resetStmtsp = nullptr;  // Statements that reset member events
+    AstCFunc* m_constructor = nullptr;  // Current class constructor
+    AstCFunc* m_destructor = nullptr;  // Current class destructor
+
+    // METHODS
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // VISITORS
+    virtual void visit(AstClass* nodep) override {
+        VL_RESTORER(m_classp);
+        VL_RESTORER(m_resetStmtsp);
+        VL_RESTORER(m_destructor);
+        m_classp = nodep;
+        iterateChildren(nodep);
+        if (m_resetStmtsp) {
+            UASSERT_OBJ(m_constructor, nodep, "Class has no constructor");
+            UASSERT_OBJ(m_destructor, nodep, "Class has no destructor");
+            nodep->addMembersp(new AstVar{nodep->fileline(), VVarType::MEMBER, "vlSymsp",
+                                          nodep->findBasicDType(VBasicDTypeKwd::SYMSPTR)});
+            m_constructor->addStmtsp(
+                new AstCStmt{nodep->fileline(), "this->vlSymsp = vlSymsp;\n"});
+            m_destructor->addStmtsp(m_resetStmtsp);
+        }
+    }
+    virtual void visit(AstVar* nodep) override {
+        if (!m_classp) return;
+        if (nodep->dtypep()->basicp() && nodep->dtypep()->basicp()->isEventValue()) {
+            m_resetStmtsp = AstNode::addNext(
+                m_resetStmtsp,
+                new AstCStmt{nodep->fileline(), "vlSymsp->__Vm_eventDispatcher.cancel("
+                                                    + nodep->nameProtect() + ");\n"});
+        }
+    }
+    virtual void visit(AstCFunc* nodep) override {
+        if (nodep->isDestructor())
+            m_destructor = nodep;
+        else if (nodep->isConstructor())
+            m_constructor = nodep;
+    }
+
+    //--------------------
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
```

Comment:
Would recommend an additional `virtual void visit(AstNodeModeule*) override { } // Accelarate` to speed things along. At that point you probably can also just assert `m_classp` is not `nullptr` in `visit(AstVar*)` above.

This is assuming other modules don't contain `AstClass`, which I think should be true.